File: backend/tools_service.py
from __future__ import annotations

import json
import re
import subprocess
import sys
from datetime import datetime, timezone
from pathlib import Path
from typing import Any, Optional

from .config import (
    DEFAULT_SKILL_SCAN_DIRS,
    FIND_SKILL_CATALOGUE,
    FIND_SKILL_INSTALL_SCRIPT,
    FIND_SKILL_ROOT,
    FIND_SKILL_UPDATE_SCRIPT,
    MCP_CONFIG_CANDIDATES,
    SKILLS_ROOT,
    TOOLS_CONFIG_PATH,
    WEBNOVEL_PY,
)


def _now() -> str:
    return datetime.now(timezone.utc).isoformat()


def _load_tools_config() -> dict[str, Any]:
    if not TOOLS_CONFIG_PATH.is_file():
        return {"skill_scan_dirs": [], "extra_paths": {}}
    return json.loads(TOOLS_CONFIG_PATH.read_text(encoding="utf-8"))


def _save_tools_config(data: dict[str, Any]) -> dict[str, Any]:
    TOOLS_CONFIG_PATH.write_text(
        json.dumps(data, ensure_ascii=False, indent=2),
        encoding="utf-8",
    )
    return data


def get_tools_overview() -> dict[str, Any]:
    cfg = _load_tools_config()
    installed = scan_installed_skills()
    catalogue = catalogue_meta()
    mcp = list_mcp_configs()
    return {
        "literary_writer": {
            "path": str(SKILLS_ROOT),
            "webnovel_cli": WEBNOVEL_PY.is_file(),
            "exists": SKILLS_ROOT.is_dir(),
        },
        "find_skill": {
            "path": str(FIND_SKILL_ROOT),
            "catalogue_available": FIND_SKILL_CATALOGUE.is_file(),
            "install_script": FIND_SKILL_INSTALL_SCRIPT.is_file(),
            **catalogue,
        },
        "installed_skills_count": len(installed),
        "mcp_servers_count": sum(len(m.get("servers") or []) for m in mcp.get("files") or []),
        "skill_scan_dirs": get_skill_scan_dirs(),
        "config": cfg,
    }


def get_skill_scan_dirs() -> list[str]:
    cfg = _load_tools_config()
    extra = [Path(p).expanduser() for p in cfg.get("skill_scan_dirs") or []]
    dirs: list[Path] = []
    seen: set[str] = set()
    for p in [*DEFAULT_SKILL_SCAN_DIRS, *extra]:
        try:
            resolved = p.resolve()
        except OSError:
            continue
        key = str(resolved)
        if key not in seen and resolved.is_dir():
            seen.add(key)
            dirs.append(resolved)
    return [str(d) for d in dirs]


def _parse_frontmatter(text: str) -> dict[str, str]:
    if not text.startswith("---"):
        return {}
    end = text.find("\n---", 3)
    if end < 0:
        return {}
    block = text[3:end].strip()
    meta: dict[str, str] = {}
    for line in block.splitlines():
        if ":" not in line:
            continue
        key, _, val = line.partition(":")
        meta[key.strip()] = val.strip().strip('"').strip("'")
    return meta


def _skill_entry(skill_md: Path, root: Path) -> dict[str, Any]:
    rel = skill_md.parent.relative_to(root)
    text = skill_md.read_text(encoding="utf-8", errors="replace")
    meta = _parse_frontmatter(text)
    name = meta.get("name") or skill_md.parent.name
    sub_skills: list[str] = []
    skills_dir = skill_md.parent / "skills"
    if skills_dir.is_dir():
        for sub in sorted(skills_dir.glob("*/SKILL.md")):
            sub_skills.append(sub.parent.name)
    return {
        "id": str(rel).replace("/", "__"),
        "name": name,
        "folder": str(skill_md.parent),
        "relative_path": str(rel),
        "description": (meta.get("description") or "")[:280],
        "has_scripts": (skill_md.parent / "scripts").is_dir(),
        "sub_skills": sub_skills,
        "is_literary_writer": "literary-writer" in skill_md.parent.name,
    }


def scan_installed_skills() -> list[dict[str, Any]]:
    items: list[dict[str, Any]] = []
    seen_folders: set[str] = set()
    for scan_root in get_skill_scan_dirs():
        root = Path(scan_root)
        for skill_md in sorted(root.glob("**/SKILL.md")):
            folder = str(skill_md.parent.resolve())
            if folder in seen_folders:
                continue
            # 跳过 find-skill 子目录里的缓存等
            if "/cache/" in str(skill_md):
                continue
            seen_folders.add(folder)
            try:
                items.append(_skill_entry(skill_md, root))
            except (OSError, ValueError):
                continue
    items.sort(key=lambda x: (not x.get("is_literary_writer"), x.get("name", "")))
    return items


def catalogue_meta() -> dict[str, Any]:
    if not FIND_SKILL_CATALOGUE.is_file():
        return {"total": 0, "updated_at": None, "stale": True}
    data = json.loads(FIND_SKILL_CATALOGUE.read_text(encoding="utf-8"))
    updated = data.get("updated_at")
    stale = True
    if updated:
        try:
            dt = datetime.fromisoformat(str(updated).replace("Z", "+00:00"))
            age_days = (datetime.now(timezone.utc) - dt).days
            stale = age_days > 30
        except ValueError:
            pass
    return {
        "total": data.get("total", len(data.get("skills") or [])),
        "updated_at": updated,
        "stale": stale,
        "sources": data.get("sources") or {},
    }


def _clean_catalogue_name(raw: str) -> str:
    """[docx](url) -> docx"""
    m = re.match(r"\[([^\]]+)\]", raw or "")
    return m.group(1) if m else (raw or "").strip()


def search_catalogue(
    query: str = "",
    *,
    limit: int = 20,
    page: int = 1,
    agent: str = "any",
) -> dict[str, Any]:
    if not FIND_SKILL_CATALOGUE.is_file():
        return {
            "items": [],
            "total": 0,
            "page": page,
            "limit": limit,
            "query": query,
            "message": "未找到 find-skill 目录，请先安装 find-skill 技能",
        }
    data = json.loads(FIND_SKILL_CATALOGUE.read_text(encoding="utf-8"))
    skills: list[dict[str, Any]] = data.get("skills") or []
    q = query.strip().lower()
    filtered: list[dict[str, Any]] = []
    for s in skills:
        agents = s.get("agents") or []
        if agent != "any" and agent not in agents:
            continue
        name = _clean_catalogue_name(str(s.get("name") or ""))
        desc = str(s.get("description") or "")
        repo = str(s.get("repo") or "")
        hay = f"{name} {desc} {repo} {' '.join(s.get('tags') or [])}".lower()
        if q and q not in hay:
            continue
        filtered.append(
            {
                "name": name,
                "raw_name": s.get("name"),
                "description": desc[:240],
                "repo": repo,
                "repo_url": s.get("repo_url"),
                "install_url": s.get("install_url"),
                "stars": s.get("stars"),
                "agents": agents,
                "source": s.get("source"),
                "category": s.get("category"),
                "install_ref": repo or name,
            }
        )
    filtered.sort(
        key=lambda x: int(str(x.get("stars") or "0").replace(",", "") or 0),
        reverse=True,
    )
    total = len(filtered)
    start = max(0, (page - 1) * limit)
    end = start + limit
    return {
        "items": filtered[start:end],
        "total": total,
        "page": page,
        "limit": limit,
        "query": query,
        "meta": catalogue_meta(),
    }


def install_skill(
    source: str,
    *,
    target: str = "claude",
    force: bool = False,
) -> dict[str, Any]:
    if not FIND_SKILL_INSTALL_SCRIPT.is_file():
        raise FileNotFoundError(f"未找到安装脚本: {FIND_SKILL_INSTALL_SCRIPT}")
    src = source.strip()
    if not src:
        raise ValueError("请提供 skill 名称或 GitHub 地址")
    cmd = [
        "bash",
        str(FIND_SKILL_INSTALL_SCRIPT),
        src,
        "--target",
        target,
    ]
    if force:
        cmd.append("--force")
    proc = subprocess.run(
        cmd,
        capture_output=True,
        text=True,
        encoding="utf-8",
        timeout=300,
    )
    if proc.returncode != 0:
        err = (proc.stderr or proc.stdout or "安装失败").strip()
        raise RuntimeError(err[:800])
    return {
        "ok": True,
        "source": src,
        "target": target,
        "output": (proc.stdout or "")[-2000:],
    }


def update_catalogue() -> dict[str, Any]:
    if not FIND_SKILL_UPDATE_SCRIPT.is_file():
        raise FileNotFoundError(f"未找到更新脚本: {FIND_SKILL_UPDATE_SCRIPT}")
    proc = subprocess.run(
        ["bash", str(FIND_SKILL_UPDATE_SCRIPT)],
        capture_output=True,
        text=True,
        encoding="utf-8",
        timeout=600,
    )
    if proc.returncode != 0:
        raise RuntimeError((proc.stderr or proc.stdout or "更新失败")[:800])
    return {"ok": True, "meta": catalogue_meta(), "output": (proc.stdout or "")[-1500:]}


MCP_TEMPLATE = """{
  "mcpServers": {
    "example-server": {
      "command": "npx",
      "args": ["-y", "@modelcontextprotocol/server-filesystem", "/path/to/allowed"],
      "env": {}
    }
  }
}
"""


def list_mcp_configs() -> dict[str, Any]:
    files: list[dict[str, Any]] = []
    for path in MCP_CONFIG_CANDIDATES:
        entry: dict[str, Any] = {
            "path": str(path),
            "exists": path.is_file(),
            "servers": [],
            "error": None,
        }
        if path.is_file():
            try:
                raw = json.loads(path.read_text(encoding="utf-8"))
                servers = raw.get("mcpServers") or raw.get("servers") or {}
                if isinstance(servers, dict):
                    for name, cfg in servers.items():
                        if not isinstance(cfg, dict):
                            continue
                        entry["servers"].append(
                            {
                                "name": name,
                                "command": cfg.get("command"),
                                "url": cfg.get("url"),
                                "args": cfg.get("args"),
                                "env_keys": list((cfg.get("env") or {}).keys()),
                            }
                        )
            except (OSError, json.JSONDecodeError) as exc:
                entry["error"] = str(exc)
        files.append(entry)
    primary = next((f for f in files if f["exists"]), files[0] if files else None)
    return {
        "files": files,
        "primary_path": primary["path"] if primary else str(MCP_CONFIG_CANDIDATES[0]),
        "template": MCP_TEMPLATE,
    }


def save_mcp_config(path: str, content: dict[str, Any]) -> dict[str, Any]:
    target = Path(path).expanduser().resolve()
    if target.name != "mcp.json":
        raise ValueError("仅允许写入 mcp.json 配置文件")
    allowed_parents = {p.parent.resolve() for p in MCP_CONFIG_CANDIDATES}
    if target.parent not in allowed_parents and target.parent != Path.home() / ".cursor":
        raise ValueError("不允许写入该路径")
    target.parent.mkdir(parents=True, exist_ok=True)
    target.write_text(json.dumps(content, ensure_ascii=False, indent=2) + "\n", encoding="utf-8")
    return {"ok": True, "path": str(target)}


def set_literary_writer_root(path: str) -> dict[str, Any]:
    import os
    import backend.config as config_mod

    p = Path(path).expanduser().resolve()
    if not p.is_dir():
        raise FileNotFoundError(f"目录不存在: {p}")
    cfg = _load_tools_config()
    cfg["literary_writer_root"] = str(p)
    _save_tools_config(cfg)
    os.environ["LITERARY_WRITER_ROOT"] = str(p)
    config_mod.SKILLS_ROOT = p
    config_mod.SCRIPTS_DIR = p / "scripts"
    config_mod.WEBNOVEL_PY = config_mod.SCRIPTS_DIR / "webnovel.py"
    return {
        "path": str(p),
        "webnovel_cli": config_mod.WEBNOVEL_PY.is_file(),
    }

File: backend-node/skill-adapter/tools-service.js
import fs from 'fs';
import path from 'path';
import os from 'os';
import { fileURLToPath } from 'url';
import { execFile } from 'child_process';
import { promisify } from 'util';
import * as sqlAdapter from '../storage/sqlite/adapter.js';
import { discoverSkillCapabilities } from './discover.js';
import {
  BUNDLED_LITERARY_WRITER,
  BUNDLED_SKILLS_DIR,
  bundledLiteraryWriterAvailable,
} from './literary-writer-paths.js';

const __dirname = path.dirname(fileURLToPath(import.meta.url));
const ROOT = path.resolve(__dirname, '..');
const DATA_DIR = path.resolve(process.env.LITERARY_STUDIO_DATA || path.join(ROOT, 'data'));
const TOOLS_CONFIG_PATH = path.join(DATA_DIR, 'tools.json');

const DEFAULT_SKILL_SCAN_DIRS = [
  ...(bundledLiteraryWriterAvailable() ? [BUNDLED_SKILLS_DIR] : []),
  path.join(os.homedir(), '.claude', 'skills'),
  path.join(os.homedir(), '.cursor', 'skills'),
  path.join(os.homedir(), '.codex', 'skills'),
];

const FIND_SKILL_ROOT = path.resolve(
  process.env.FIND_SKILL_ROOT || path.join(os.homedir(), '.claude', 'skills', 'find-skill'),
);
const FIND_SKILL_CATALOGUE = path.join(FIND_SKILL_ROOT, 'cache', 'catalogue.json');
const FIND_SKILL_INSTALL_SCRIPT = path.join(FIND_SKILL_ROOT, 'scripts', 'install-skill.sh');
const FIND_SKILL_UPDATE_SCRIPT = path.join(FIND_SKILL_ROOT, 'update-skills-catalogue.sh');
const execFileAsync = promisify(execFile);

async function runBashScript(scriptPath, args = [], timeoutMs = 300000) {
  const { stdout } = await execFileAsync('bash', [scriptPath, ...args], {
    encoding: 'utf-8',
    timeout: timeoutMs,
    maxBuffer: 4 * 1024 * 1024,
  });
  return stdout;
}

export function loadToolsConfig() {
  try {
    const fromKv = sqlAdapter.useSqlite()
      ? sqlAdapter.readKv(sqlAdapter.KV_KEYS.tools, TOOLS_CONFIG_PATH, null, (p, fb) => {
        try { return JSON.parse(fs.readFileSync(p, 'utf-8')); } catch { return fb; }
      })
      : null;
    if (fromKv) return fromKv;
    if (fs.existsSync(TOOLS_CONFIG_PATH)) {
      return JSON.parse(fs.readFileSync(TOOLS_CONFIG_PATH, 'utf-8'));
    }
  } catch {}
  return { skill_scan_dirs: [], extra_paths: {} };
}

export function saveToolsConfig(data) {
  fs.mkdirSync(DATA_DIR, { recursive: true });
  fs.writeFileSync(TOOLS_CONFIG_PATH, JSON.stringify(data, null, 2), 'utf-8');
  if (sqlAdapter.useSqlite()) {
    sqlAdapter.writeKv(sqlAdapter.KV_KEYS.tools, TOOLS_CONFIG_PATH, data, (p, d) => {
      fs.writeFileSync(p, JSON.stringify(d, null, 2), 'utf-8');
    });
  }
  return data;
}

function resolveLiteraryWriterRoot() {
  const cfg = loadToolsConfig();
  const custom = String(cfg.literary_writer_root || '').trim();
  if (custom) {
    const p = path.resolve(custom.replace(/^~/, os.homedir()));
    if (fs.existsSync(p) && fs.statSync(p).isDirectory()) return p;
  }
  const env = process.env.LITERARY_WRITER_ROOT;
  if (env) {
    const p = path.resolve(env.replace(/^~/, os.homedir()));
    if (fs.existsSync(p) && fs.statSync(p).isDirectory()) return p;
  }
  const candidates = [
    BUNDLED_LITERARY_WRITER,
    path.join(os.homedir(), '.claude', 'skills', 'literary-writer'),
    path.join(os.homedir(), '.cursor', 'skills', 'literary-writer'),
    path.join(ROOT, '..', 'skills', 'literary-writer'),
  ];
  for (const c of candidates) {
    const resolved = path.resolve(c);
    if (fs.existsSync(resolved) && fs.statSync(resolved).isDirectory()) {
      return resolved;
    }
  }
  return path.resolve(candidates[0]);
}

function webnovelPyPath(skillsRoot) {
  return path.join(skillsRoot, 'scripts', 'webnovel.py');
}

function getSkillScanDirs() {
  const cfg = loadToolsConfig();
  const extra = (cfg.skill_scan_dirs || []).map((p) =>
    path.resolve(String(p).replace(/^~/, os.homedir())),
  );
  const dirs = [];
  const seen = new Set();
  for (const p of [...DEFAULT_SKILL_SCAN_DIRS, ...extra]) {
    try {
      if (!fs.existsSync(p) || !fs.statSync(p).isDirectory()) continue;
      const key = fs.realpathSync(p);
      if (seen.has(key)) continue;
      seen.add(key);
      dirs.push(key);
    } catch {}
  }
  return dirs;
}

function parseFrontmatter(text) {
  if (!text.startsWith('---')) return {};
  const end = text.indexOf('\n---', 3);
  if (end < 0) return {};
  const block = text.slice(3, end).trim();
  const meta = {};
  for (const line of block.split('\n')) {
    const i = line.indexOf(':');
    if (i < 0) continue;
    const key = line.slice(0, i).trim();
    let val = line.slice(i + 1).trim();
    val = val.replace(/^["']|["']$/g, '');
    meta[key] = val;
  }
  return meta;
}

function skillEntry(skillMd, scanRoot) {
  const folder = path.dirname(skillMd);
  const text = fs.readFileSync(skillMd, 'utf-8');
  const meta = parseFrontmatter(text);
  const rel = path.relative(scanRoot, folder);
  const subSkills = [];
  const skillsDir = path.join(folder, 'skills');
  if (fs.existsSync(skillsDir)) {
    for (const name of fs.readdirSync(skillsDir)) {
      const sub = path.join(skillsDir, name, 'SKILL.md');
      if (fs.existsSync(sub)) subSkills.push(name);
    }
  }
  let entry_relative = null;
  let command_count = 0;
  try {
    const caps = discoverSkillCapabilities(folder);
    entry_relative = caps.entry?.relative || null;
    command_count = Object.keys(caps.commands || {}).length;
  } catch {}
  return {
    id: rel.split(path.sep).join('__'),
    name: meta.name || path.basename(folder),
    folder,
    relative_path: rel,
    description: (meta.description || '').slice(0, 280),
    has_scripts: fs.existsSync(path.join(folder, 'scripts')),
    sub_skills: subSkills.sort(),
    is_literary_writer: folder.includes('literary-writer'),
    invocable: Boolean(entry_relative),
    entry_relative,
    command_count,
  };
}

export function scanInstalledSkills() {
  const items = [];
  const seen = new Set();
  for (const scanRoot of getSkillScanDirs()) {
    walkSkillMd(scanRoot, scanRoot, (skillMd) => {
      const folder = fs.realpathSync(path.dirname(skillMd));
      if (seen.has(folder)) return;
      if (skillMd.includes(`${path.sep}cache${path.sep}`)) return;
      seen.add(folder);
      try {
        items.push(skillEntry(skillMd, scanRoot));
      } catch {}
    });
  }
  items.sort((a, b) => {
    if (a.is_literary_writer !== b.is_literary_writer) {
      return a.is_literary_writer ? -1 : 1;
    }
    return (a.name || '').localeCompare(b.name || '');
  });
  return items;
}

/**
 * Fast counter used by overview/status endpoints.
 * Limits traversal work to avoid blocking packaged (DMG) builds on large local skill trees.
 */
function countInstalledSkillsFast({ maxMs = 1500, maxDirs = 5000 } = {}) {
  const deadline = Date.now() + maxMs;
  const seenFolders = new Set();
  let count = 0;
  let visitedDirs = 0;
  let truncated = false;

  const walk = (dir) => {
    if (Date.now() > deadline || visitedDirs > maxDirs) {
      truncated = true;
      return;
    }
    visitedDirs += 1;
    let entries;
    try {
      entries = fs.readdirSync(dir, { withFileTypes: true });
    } catch {
      return;
    }
    for (const ent of entries) {
      if (Date.now() > deadline || visitedDirs > maxDirs) {
        truncated = true;
        return;
      }
      const full = path.join(dir, ent.name);
      if (ent.isDirectory()) {
        walk(full);
        continue;
      }
      if (ent.name !== 'SKILL.md' || full.includes(`${path.sep}cache${path.sep}`)) continue;
      let folder;
      try {
        folder = fs.realpathSync(path.dirname(full));
      } catch {
        continue;
      }
      if (seenFolders.has(folder)) continue;
      seenFolders.add(folder);
      count += 1;
    }
  };

  for (const scanRoot of getSkillScanDirs()) {
    walk(scanRoot);
    if (truncated) break;
  }

  return { count, truncated };
}

function walkSkillMd(dir, scanRoot, onFile) {
  let entries;
  try {
    entries = fs.readdirSync(dir, { withFileTypes: true });
  } catch {
    return;
  }
  for (const ent of entries) {
    const full = path.join(dir, ent.name);
    if (ent.isDirectory()) {
      walkSkillMd(full, scanRoot, onFile);
    } else if (ent.name === 'SKILL.md') {
      onFile(full);
    }
  }
}

function catalogueMeta() {
  if (!fs.existsSync(FIND_SKILL_CATALOGUE)) {
    return { total: 0, updated_at: null, stale: true };
  }
  try {
    const data = JSON.parse(fs.readFileSync(FIND_SKILL_CATALOGUE, 'utf-8'));
    const updated = data.updated_at;
    let stale = true;
    if (updated) {
      const dt = new Date(updated);
      const ageDays = (Date.now() - dt.getTime()) / (86400 * 1000);
      stale = ageDays > 30;
    }
    return {
      total: data.total ?? (data.skills || []).length,
      updated_at: updated,
      stale,
      sources: data.sources || {},
    };
  } catch {
    return { total: 0, updated_at: null, stale: true };
  }
}

export function getToolsOverview() {
  const skillsRoot = resolveLiteraryWriterRoot();
  const webnovel = webnovelPyPath(skillsRoot);
  const installed = countInstalledSkillsFast();
  return {
    literary_writer: {
      path: skillsRoot,
      webnovel_cli: fs.existsSync(webnovel),
      webnovel_path: webnovel,
      exists: fs.existsSync(skillsRoot),
    },
    find_skill: {
      path: FIND_SKILL_ROOT,
      catalogue_available: fs.existsSync(FIND_SKILL_CATALOGUE),
      install_script: fs.existsSync(FIND_SKILL_INSTALL_SCRIPT),
      ...catalogueMeta(),
    },
    installed_skills_count: installed.count,
    installed_skills_count_partial: installed.truncated,
    skill_scan_dirs: getSkillScanDirs(),
    config: loadToolsConfig(),
  };
}

export function setLiteraryWriterRoot(dirPath) {
  const p = path.resolve(dirPath.replace(/^~/, os.homedir()));
  if (!fs.existsSync(p) || !fs.statSync(p).isDirectory()) {
    throw new Error(`目录不存在: ${p}`);
  }
  const cfg = loadToolsConfig();
  cfg.literary_writer_root = p;
  saveToolsConfig(cfg);
  process.env.LITERARY_WRITER_ROOT = p;
  const webnovel = webnovelPyPath(p);
  return {
    path: p,
    webnovel_cli: fs.existsSync(webnovel),
    webnovel_path: webnovel,
  };
}

export function searchCatalogue(query = '', { limit = 20, page = 1, agent = 'any' } = {}) {
  if (!fs.existsSync(FIND_SKILL_CATALOGUE)) {
    return {
      items: [],
      total: 0,
      page,
      limit,
      query,
      message: '未找到 find-skill 目录，请先安装 find-skill 技能',
    };
  }
  const data = JSON.parse(fs.readFileSync(FIND_SKILL_CATALOGUE, 'utf-8'));
  let skills = data.skills || [];
  const q = query.trim().toLowerCase();
  if (q) {
    skills = skills.filter((s) => {
      const agents = s.agents || [];
      if (agent !== 'any' && !agents.includes(agent)) return false;
      const name = String(s.name || '').replace(/^\[([^\]]+)\].*/, '$1');
      const hay = `${name} ${s.description || ''} ${s.repo || ''} ${(s.tags || []).join(' ')}`.toLowerCase();
      return hay.includes(q);
    });
  }
  const total = skills.length;
  const start = (page - 1) * limit;
  const items = skills.slice(start, start + limit);
  return { items, total, page, limit, query };
}

let catalogueUpdateInFlight = null;
let installSkillInFlight = null;

export async function installSkill({ source, target = 'claude', force = false }) {
  if (!fs.existsSync(FIND_SKILL_INSTALL_SCRIPT)) {
    throw new Error(`未找到安装脚本: ${FIND_SKILL_INSTALL_SCRIPT}`);
  }
  if (installSkillInFlight) return installSkillInFlight;
  installSkillInFlight = (async () => {
    try {
      const args = [source, '--target', target];
      if (force) args.push('--force');
      const out = await runBashScript(FIND_SKILL_INSTALL_SCRIPT, args, 300000);
      return { ok: true, source, target, output: out.slice(-2000) };
    } finally {
      installSkillInFlight = null;
    }
  })();
  return installSkillInFlight;
}

export async function updateCatalogue() {
  if (!fs.existsSync(FIND_SKILL_UPDATE_SCRIPT)) {
    throw new Error(`未找到更新脚本: ${FIND_SKILL_UPDATE_SCRIPT}`);
  }
  if (catalogueUpdateInFlight) return catalogueUpdateInFlight;
  catalogueUpdateInFlight = (async () => {
    try {
      const out = await runBashScript(FIND_SKILL_UPDATE_SCRIPT, [], 600000);
      return { ok: true, meta: catalogueMeta(), output: out.slice(-1500) };
    } finally {
      catalogueUpdateInFlight = null;
    }
  })();
  return catalogueUpdateInFlight;
}

